Enhance Run a Mithril Signer as an SPO guide (#1071)
input-output-hk/mithril: docs/root/manual/getting-started/run-signer-node.md

Reviewer: You forgot to put an example for the squid configuration.
Also watch out as it will change probably before you merge with this PR #1089

- On `mainnet`, you **must** run the **production** deployment where the **Mithril Signer** runs on the **Cardano block producer** machine and the **Mithril relay** runs on the **Cardano relay** machine. **Note** that you can run the **production** deployment on `testnet`.
- You can also run **naive** deployment, where the **Mithril Signer** runs on the **Cardano relay** machine. This is possible in the testnet environment only, and does not require setting up a **Mithril relay**.

> In the current setup, you don't need to install a Mithril Aggregator.

Reviewer: You could put that sentence in a separate tip block

  * `RELAY_ENDPOINT=http://192.168.1.50:3128` **(optional)**: this is the endpoint of the **Mithril relay**, which is required for **production** deployment only. For **naive** deployment, do not set this variable in your environment file.
:::

:::tip

> You can find below the `ERA_READER_ADAPTER_TYPE` and `ERA_READER_ADAPTER_PARAMS` values for **release-preprod**. Other values must be replaced as explained in the caution box above.

Reviewer: ```suggestion
Here is an **example** set of values for **release-preprod** that will be used in this guide in the **tip** boxes to illustrate some commands.
```

* In the `/opt/mithril/mithril-signer/service.env` env file:
  * `KES_SECRET_KEY_PATH=/cardano/keys/kes.skey`
  * `OPERATIONAL_CERTIFICATE_PATH=/cardano/cert/opcert.cert`
  * `DB_DIRECTORY=/cardano/db`
  * `CARDANO_NODE_SOCKET_PATH=/cardano/ipc/node.socket`
  * `CARDANO_CLI_PATH=/app/bin/cardano-cli`
  * `DATA_STORES_DIRECTORY=/opt/mithril/stores`
  * `STORE_RETENTION_LIMIT=5`
  * `ERA_READER_ADAPTER_TYPE=cardano-chain`
  * `ERA_READER_ADAPTER_PARAMS={"address": "addr_test1qpkyv2ws0deszm67t840sdnruqgr492n80g3y96xw3p2ksk6suj5musy6w8lsg3yjd09cnpgctc2qh386rtxphxt248qr0npnx", "verification_key": "5b35352c3232382c3134342c38372c3133382c3133362c34382c382c31342c3138372c38352c3134382c39372c3233322c3235352c3232392c33382c3234342c3234372c3230342c3139382c31332c33312c3232322c32352c3136342c35322c3130322c39312c3132302c3230382c3134375d"}`

>   * `RELAY_ENDPOINT=http://192.168.1.50:3128`

Reviewer: Just list the values associated to each value placeholder encapsulated in `**`

  * `RELAY_ENDPOINT=http://192.168.1.50:3128` **(optional)**: this is the endpoint of the **Mithril relay**, which is required for **production** deployment only. For **naive** deployment, do not set this variable in your environment file.
:::

:::tip

> Here is an **example** set of values for **release-preprod** that will be used in this guide in the **tip** boxes to illustrate some commands.  

Reviewer: ```suggestion
Here is an **example** set of values for **release-preprod** that will be used in this guide in the **tip** boxes to illustrate some commands:  
```

* ****YOUR_KES_SECRET_KEY_PATH****: `/cardano/keys/kes.skey`
* ****YOUR_OPERATIONAL_CERTIFICATE_PATH****: `/cardano/keys/node.cert`
* ****YOUR_CARDANO_NETWORK****: `preprod`
* ****YOUR_AGGREGATOR_ENDPOINT****: `https://aggregator.release-preprod.api.mithril.network/aggregator`
* ****YOUR_ERA_READER_ADAPTER_TYPE****: `cardano-chain`
* ****YOUR_ERA_READER_ADAPTER_PARAMS****: `{"address": "addr_test1qpkyv2ws0deszm67t840sdnruqgr492n80g3y96xw3p2ksk6suj5musy6w8lsg3yjd09cnpgctc2qh386rtxphxt248qr0npnx", "verification_key": "5b35352c3232382c3134342c38372c3133382c3133362c34382c382c31342c3138372c38352c3134382c39372c3233322c3235352c3232392c33382c3234342c3234372c3230342c3139382c31332c33312c3232322c32352c3136342c35322c3130322c39312c3132302c3230382c3134375d"}`
* ****YOUR_RELAY_ENDPOINT****: `192.168.1.50`
* ****YOUR_RELAY_LISTENING_PORT**** = `3128`
* ****YOUR_BLOCK_PRODUCER_INTERNAL_IP****: `192.168.1.75`
* ****YOUR_SIGNER_LOGS_PATH****: `/var/log/syslog`

> * ****YOUR_PARTY_ID****: `pool1hp72sauk0g0yqm4dzllz0pz6j93gewhllkzphn4hykkfmne43y`

Author: ![image](https://github.com/input-output-hk/mithril/assets/135982616/18c3b013-fb22-4270-8062-7f32e5a5fa05)

Is it good like that ?

EOF'
```

:::tip

> Here is an example command created with the example set for `release-preprod`:

Reviewer: Can you do this modification here and in all the other occurrences?
```suggestion
Here is an example of the aforementioned command created with the example set for `release-preprod`:
```

EOF'
```

:::tip

> Here is an example of the env file for **release-preprod** :

Reviewer: ```suggestion
Here is an example command created with the example set for `release-preprod`:
```

> Here is an example of the env file for **release-preprod** :

Reviewer: We should maybe have the same for the naive deployment

sudo ufw allow from **YOUR_BLOCK_PRODUCER_INTERNAL_IP** to any port **YOUR_RELAY_LISTENING_PORT** proto tcp
```

:::tip

> Here is a example of the command to run :

Reviewer: You can use this for all the other tip boxes
```suggestion
Here is an example command created with the example set for `release-preprod`:
```

```bash

> sudo ufw allow from 123.123.123.123 to any port 3128 proto tcp

Reviewer: `123.123.123.123` should be a real IP that you define in the set of parameters above

sudo service netfilter-persistent save
```

:::tip
Here is a example of the command to run :

```bash

> sudo iptables -A INPUT -s 123.123.123.123 -p tcp --dport 3128 -j ACCEPT

Reviewer: Same here

sudo iptables -L -v
sudo service netfilter-persistent save
```
:::

## Verify the Mithril Signer deployment

:::tip
You can verify that your signer node is correctly configured and registered by checking your signer node logs or by running this command :

```bash
if grep -q "STATE MACHINE: new cycle: Registered" **YOUR_SIGNER_NODE_LOGS_PATH**; then
    echo ">> Your signer node is registered"
else
    echo ">> Your signer node is not registered"
fi
```

> 

Reviewer: This should be a separate section below the existing tip box:
### Verify that your signer is registered

Also:
- The command should probably first write to a file, chmod the file and the use it for clarity
- Add a tip box for the associated example command

Once this delay has passed, you should be able to observe your `PoolId` listed in some of the certificates accessible on the [`Mithril Explorer`](https://mithril.network/explorer) or by running the following command : 

```bash
party_id=**YOUR_POOL_ID**
aggregator_endpoint=**YOUR_AGGREGATOR_ENDPOINT**
found=false

curl -s "$aggregator_endpoint/certificates" -H 'accept: application/json' | jq -r '.[] | .hash' | while read -r hash; do
    response=$(curl -s "$aggregator_endpoint/certificate/$hash" -H 'accept: application/json')
    signer_count=$(echo "$response" | jq '.metadata.signers | length')
    for (( i=0; i < signer_count; i++ )); do
        party_id_response=$(echo "$response" | jq -r ".metadata.signers[$i].party_id")
        if [[ "$party_id_response" == "$party_id" ]]; then
            echo ">> You have signed this certificate : $hash"
            found=true
            break
        fi
    done

    if [ "$found" = true ]; then
        break
    fi
done

if [ "$found" = false ]; then
    echo ">> Your party id was not found in the last certificates"

> fi

Reviewer: Same here, create a file and explain how to run it.
Alos the env vars should be capitalized

EOF'
```

:::tip

Here is an example command created with the example set for `release-preprod`:

```bash
sudo bash -c 'cat > /etc/squid/squid.conf << EOF
http_port 3128
acl relay_internal_ip src 192.168.1.75
acl aggregator_domain dstdomain .mithril.network
acl SSL_port port 443
http_access allow relay_internal_ip aggregator_domain SSL_port
forwarded_for delete
via off
follow_x_forwarded_for deny all
request_header_access Authorization allow all
request_header_access Proxy-Authorization allow all
request_header_access Cache-Control allow all
request_header_access Content-Length allow all
request_header_access Content-Type allow all
request_header_access Date allow all
request_header_access Host allow all
request_header_access If-Modified-Since allow all
request_header_access Pragma allow all
request_header_access Accept allow all
request_header_access Accept-Charset allow all
request_header_access Accept-Encoding allow all
request_header_access Accept-Language allow all
request_header_access Connection allow all
request_header_access All deny all
cache deny all

> http_access deny all

Reviewer: All the comments and carriage returns have been stripped :sweat_smile:

sudo service netfilter-persistent save
```

:::tip

Here is an example command created with the example set for `release-preprod`:

```bash
sudo iptables -A INPUT -s 192.168.1.75 -p tcp --dport 3128 -j ACCEPT
sudo iptables -L -v
sudo service netfilter-persistent save
```

:::

## Verify the Mithril Signer deployment

:::tip
There is a delay of `2` epochs between the registration of the signer node and its ability to generate individual signatures. This delay is further explained in the [Mithril certificate chain in depth](https://mithril.network/doc/mithril/mithril-protocol/certificates) documentation.

Once this delay has passed, you should be able to observe your `PoolId` listed in some of the certificates accessible on the [`Mithril Explorer`](https://mithril.network/explorer).
:::

> ### Verify your Signer is registered

Reviewer: `### Verify your signer is registered`
Also in SPO onboarding guide :slightly_smiling_face:

Author: This is not what has been achieved in this commit ? https://github.com/input-output-hk/mithril/pull/1071/commits/969715b050339939652f242d63813dac251a8eb7

After installing the Mithril Signer, you can verify that your node is registered by checking your Mithril Signer node logs.  

First, download the script in the desired directory:

```bash

> wget "https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-signer/scripts/verify_signer_registration.sh"

Reviewer: No need for quote:
```suggestion
wget https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-signer/scripts/verify_signer_registration.sh
```

```

Make the script executable:

```bash
chmod +x verify_signer_registration.sh
```

Then, add the environment variable corresponding to the path of your Mithril Signer node logs file:
```bash
export SIGNER_LOGS_PATH=**YOUR_SIGNER_LOGS_PATH**
```

Finally, execute the script:
```bash
./verify_signer_registration.sh
```

> ### Verify your Signer generates individual signatures

Reviewer: `### Verify your signer contributes with individual signatures`
Also in SPO onboarding guide :slightly_smiling_face:

> After waiting for the delay of 2 epochs, you will be able to verify the generation of individual signatures.

Reviewer: ```suggestion
After waiting for two epochs, you will be able to verify that your signer is contributing with individual signatures.
```

First, download the script in the desired directory:

```bash

> wget "https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-signer/scripts/verify_signer_signature.sh"

Reviewer: Same as above:
```suggestion
wget https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-signer/scripts/verify_signer_signature.sh
```

```

Make the script executable:

```bash
chmod +x verify_signer_signature.sh
```

Then, add the environment variables:
```bash

> export PARTY_ID=**YOUR_PARTY_ID**

Reviewer: Same comment as above :slightly_smiling_face:

sudo service netfilter-persistent save
```

:::tip

Here is an example command created with the example set for `release-preprod`:

```bash
sudo iptables -A INPUT -s 192.168.1.75 -p tcp --dport 3128 -j ACCEPT
sudo iptables -L -v
sudo service netfilter-persistent save
```

:::

## Verify the Mithril Signer deployment

:::tip
There is a delay of `2` epochs between the registration of the signer node and its ability to generate individual signatures. This delay is further explained in the [Mithril certificate chain in depth](https://mithril.network/doc/mithril/mithril-protocol/certificates) documentation.

Once this delay has passed, you should be able to observe your `PoolId` listed in some of the certificates accessible on the [`Mithril Explorer`](https://mithril.network/explorer).
:::

> ### Verify your signer is registered

Reviewer: Can you add an examples of the information displayed when registered and not registered at the end of the section?

Author: **_Result in the explorer :_**

![image](https://github.com/input-output-hk/mithril/assets/135982616/947114a1-159a-4beb-8479-50f46ce94c39)

Is this ok for you ?

Reviewer: Looks good!

After installing the Mithril Signer, you can verify that your node is registered by checking your Mithril Signer node logs.  

First, download the script in the desired directory:

```bash

> wget "https://raw.githubusercontent.com/input-output-hk/mithril/main/docs/scripts/verify_signer_registration.sh"

Reviewer: - The URL should be on the doc website `mithril.network`
- Also no need to add double quotes around the URL

Author: For the first point, I'm not sure to get it. The URLs should be : 
https://mithril.network/doc/scripts/verify_signer_registration.sh and https://mithril.network/doc/scripts/verify_signer_signature.sh ?

Reviewer: I think the scripts should probably be put in the the `static` folder :thinking: 

Could you do the following:
- move them in `static/scripts` folder
- verify that you can download locally the scripts at `http://localhost:3000/doc/scripts/verify_signer_registration.sh` and `http://localhost:3000/doc/scripts/verify_signer_signature.sh`
- replace the url with `https://mithril.network/doc/scripts/verify_signer_registration.sh` and `https://mithril.network/doc/scripts/verify_signer_signature.sh`

```

Make the script executable:

```bash
chmod +x verify_signer_registration.sh
```

Finally, execute the script:
```bash
SIGNER_LOGS_PATH=**YOUR_SIGNER_LOGS_PATH** ./verify_signer_registration.sh
```

:::tip

Here is an example command:

```bash
SIGNER_LOGS_PATH=/var/log/syslog ./verify_signer_registration.sh
```

:::

> ### Verify your signer contributes with individual signatures

Reviewer: Can you add an examples of the information displayed when contributing and not contributing at the end of the section?

Author: **_Result in the explorer :_**

![image](https://github.com/input-output-hk/mithril/assets/135982616/54ad2616-845f-44e6-a3df-c219b067c022)

For information, the hash mentioned at the end of the url is a fake hash.
```bash
>> Congrats, you have signed this certificate: https://aggregator.release-preprod.api.mithril.network/aggregator/certificate/el3p289b03a223244285b2ls10839846ae7a69f1e8362824a383f376f93f723f !
```

Is this ok for you ?

Reviewer: Great!

sudo service netfilter-persistent save
```

:::tip

Here is an example of the aforementioned command created with the example set for `release-preprod`:

```bash
sudo iptables -A INPUT -s 192.168.1.75 -p tcp --dport 3128 -j ACCEPT
sudo iptables -L -v
sudo service netfilter-persistent save
```

:::

## Verify the Mithril Signer deployment

:::tip
There is a delay of `2` epochs between the registration of the signer node and its ability to generate individual signatures. This delay is further explained in the [Mithril certificate chain in depth](https://mithril.network/doc/mithril/mithril-protocol/certificates) documentation.

Once this delay has passed, you should be able to observe your `PoolId` listed in some of the certificates accessible on the [`Mithril Explorer`](https://mithril.network/explorer).
:::

### Verify your signer is registered

After installing the Mithril Signer, you can verify that your node is registered by checking your Mithril Signer node logs.  

> First, download the script in the desired directory:

Reviewer 2: Should it be 'into' the desired directory? Currently, it feels like we need to say 'from', but that does not seem to be right. If you choose to update this, please consider other similar instructions below.

Author: Thanks, 'into' fits better indeed.

```bash
wget https://mithril.network/doc/scripts/verify_signer_registration.sh
```

Make the script executable:

```bash
chmod +x verify_signer_registration.sh
```

Finally, execute the script:
```bash
SIGNER_LOGS_PATH=**YOUR_SIGNER_LOGS_PATH** ./verify_signer_registration.sh
```

:::tip

Here is an example command:

```bash
SIGNER_LOGS_PATH=/var/log/syslog ./verify_signer_registration.sh
```

:::

> If you signer is registered, you should see this message:

Reviewer 2: ```suggestion
If your signer is registered, you should see this message:
```

Author: Done!
